{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import glob\n",
    "import scipy\n",
    "import pickle\n",
    "import matplotlib\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from collections import defaultdict\n",
    "from matplotlib import pyplot as plt\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "\n",
    "%matplotlib inline\n",
    "sns.set_style('whitegrid')\n",
    "pd.set_option('display.max_rows', 100)\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "pd.set_option('display.max_columns', 100)\n",
    "\n",
    "sys.path.append('/home/mattolm/Bio_scripts/')\n",
    "\n",
    "from IPython.display import display, HTML"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load info on genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "HIdb = pd.read_pickle('/data1/bio_db/refseq/analysis/singleCopyGenes_2/data_tables/Essential_RP_hmm_descriptions_v1.pickle')\n",
    "\n",
    "Ddb = pd.read_pickle('/data1/bio_db/refseq/analysis/singleCopyGenes_2/data_tables/RibosomalDistanceMatricies_v2.2.pickle')\n",
    "Ddb = Ddb[Ddb['genome1'] != Ddb['genome2']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "#tdb = Ddb[(Ddb['hmm_name'] == 'GrpE') & (Ddb['genomeList'] == 'infantList')]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Figure out information on genome sets\n",
    "\n",
    "Need to know how many true within comparisons there are and how many true without comparisons there are"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# def calculate_comps(cdb):\n",
    "#     '''\n",
    "#     Based on a cdb figure out how many within-cluster comparisons there are\n",
    "#     '''\n",
    "#     comps = 0\n",
    "#     for c, db in cdb.groupby('cluster'):\n",
    "#         comps += ((len(db) ** 2) - len(db))\n",
    "#     return comps, ((len(cdb) ** 2) - len(cdb)) - comps\n",
    "\n",
    "# table = defaultdict(list)\n",
    "# for fi in glob.glob('/data1/bio_db/refseq/analysis/MAGlists_2/dataframes/*_Cdb.csv'):\n",
    "#     li = os.path.basename(fi).split('_')[0]\n",
    "#     db = pd.read_csv(fi)\n",
    "#     within_comps, without_comps = calculate_comps(db)\n",
    "#     table['list'].append(li)\n",
    "#     table['within'].append(within_comps)\n",
    "#     table['without'].append(without_comps)\n",
    "# cdb = pd.DataFrame(table)\n",
    "# cdb"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculate thresholds"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_thresholds_3(Odb, m, values, within, without, col='ani', MIN_VAL=0, gene='test', gl='test'):\n",
    "    '''\n",
    "    Calculate thresholds based on distance matrix.\n",
    "    \n",
    "    Version 2 ignores 'unk'\n",
    "    \n",
    "    Args:\n",
    "        Odb: comparison matrix\n",
    "        cols: columns of the matrix to run on with ID values\n",
    "        metrics: columns of the matrix that are either True/False\n",
    "        values: threshold values to use\n",
    "    ''' \n",
    "    within = len(Odb[(Odb[m] == True)])\n",
    "    without = len(Odb[(Odb[m] == False)])\n",
    "    \n",
    "    if without == 0:\n",
    "        return\n",
    "    \n",
    "    for v in values:\n",
    "        db = Odb[Odb[col] >= v]\n",
    "        \n",
    "        if len(db) == 0:\n",
    "            precision = np.nan\n",
    "        else: \n",
    "            precision = len(db[db[m] == True]) / len(db)\n",
    "\n",
    "        recall = len(db[db[m] == True]) / within\n",
    "        fpr = len(db[db[m] == False]) / without\n",
    "\n",
    "        table['genomeList'].append(gl)\n",
    "        table['gene'].append(gene)\n",
    "        table['threshold'].append(v)\n",
    "        table['method'].append(col)\n",
    "        table['metric'].append(m)\n",
    "        table['precision'].append(precision)\n",
    "        table['recall'].append(recall)\n",
    "        table['fpr'].append(fpr)\n",
    "        table['tpr'].append(recall)\n",
    "\n",
    "# TEST\n",
    "# table = defaultdict(list)\n",
    "# values = list(np.arange(0.9,1.001,.001))\n",
    "\n",
    "# within_comps = cdb[cdb['list'] == 'infantList']['within'].tolist()[0]\n",
    "# without_comps = cdb[cdb['list'] == 'infantList']['without'].tolist()[0]\n",
    "\n",
    "# calculate_thresholds_3(tdb, 'same_cluster', values, within_comps, without_comps, col='ani')\n",
    "\n",
    "# pd.DataFrame(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GrpE\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Methyltransf_5\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "PGK\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L10\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L23\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L3\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L4\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L5\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_L6\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_S11\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_S13\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_S17\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_S8\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "Ribosomal_S9\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00001\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00002\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00009\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00012\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00019\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00029\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00043\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00059\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00060\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00061\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00062\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00064\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00082\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00086\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00092\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00115\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00116\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00152\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00158\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00165\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00166\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00168\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00234\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00337\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00344\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00362\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00388\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00389\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00392\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00396\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00408\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00409\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00414\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00418\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00420\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00422\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00435\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00436\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00442\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00459\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00460\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00468\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00471\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00472\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00487\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00496\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00575\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00631\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00663\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00775\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00810\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00855\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00922\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00952\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00959\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00963\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00964\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00967\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR00981\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01009\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01011\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01017\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01021\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01024\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01029\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01030\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01031\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01032\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01044\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01049\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01050\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01059\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01063\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01066\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01067\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01071\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01079\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01164\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01169\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01171\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01391\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01393\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01632\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR01953\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02012\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02013\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02027\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02191\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02350\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02386\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02387\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02397\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02432\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR02729\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR03263\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "TIGR03594\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n",
      "tRNA-synt_1d\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n"
     ]
    }
   ],
   "source": [
    "table = defaultdict(list)\n",
    "values = list(np.arange(0.9,1.001,.001))\n",
    "\n",
    "for gene, gdb in Ddb.groupby('hmm_name'):\n",
    "    print(gene)\n",
    "    for li, db in gdb.groupby('genomeList'):\n",
    "        print(li)\n",
    "        calculate_thresholds_3(db, 'same_cluster', values, within_comps, without_comps, col='ani', gene=gene, gl=li)\n",
    "        \n",
    "TRdb = pd.DataFrame(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TRdb.to_pickle('/data1/bio_db/refseq/analysis/singleCopyGenes_2/data_tables/Thresholds_v2.2.pickle')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "Ddb = pd.read_pickle('/data1/bio_db/refseq/analysis/singleCopyGenes_2/data_tables/16S_DistanceMatricies_v2.2.pickle')\n",
    "Ddb['genome1'] = Ddb['genome1'].astype(str)\n",
    "Ddb['genome2'] = Ddb['genome2'].astype(str)\n",
    "Ddb = Ddb[Ddb['genome1'] != Ddb['genome2']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16S\n",
      "infantList\n",
      "oceanList\n",
      "refseqTheta\n",
      "soilList\n"
     ]
    }
   ],
   "source": [
    "table = defaultdict(list)\n",
    "values = list(np.arange(0.9,1.001,.001))\n",
    "\n",
    "for gene, gdb in Ddb.groupby('hmm_name'):\n",
    "    print(gene)\n",
    "    for li, db in gdb.groupby('genomeList'):\n",
    "        print(li)\n",
    "        calculate_thresholds_3(db, 'same_cluster', values, within_comps, without_comps, col='ani', gene=gene, gl=li)\n",
    "        \n",
    "TR16db = pd.DataFrame(table)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "TR3db = pd.concat([TRdb, TR16db])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "#TR3db.to_pickle('/data1/bio_db/refseq/analysis/singleCopyGenes_2/data_tables/Thresholds_v3.pickle')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.6.1",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
